core/store/src/trie/mod.rs

```diff
@@ -309,6 +315,47 @@ impl RcTrieNode {
     }
 }
 
+pub trait TrieStorage: Downcast + Send + Sync {
+    fn retrieve_raw_bytes(&self, hash: &CryptoHash) -> Option<(Vec<u8>)>;
+
+    fn retrieve_rc(&self, hash: &CryptoHash) -> Option<u32>;
```

Comment:
We are trying to reduce the amount of external dependencies in runtime code, especially those that are not well-known. I suggest we replace usage of `downcast_rs` with the following code:
```
pub trait TrieStorage: Send + Sync {
    fn retrieve_raw_bytes(&self, hash: &CryptoHash) -> Option<(Vec<u8>)>;

    fn retrieve_rc(&self, hash: &CryptoHash) -> Option<u32>;

    fn as_caching_storage(&self) -> Option<&TrieCachingStorage> { None }
    fn as_recording_storage(&self) -> Option<&TrieRecordingStorage> { None }
}
impl TrieStorage for TrieRecordingStorage {
fn as_recording_storage(&self) -> Option<&TrieRecordingStorage> { Some(self) }
}
impl TrieStorage for TrieCachingStorage {
fn as_caching_storage(&self) -> Option<&TrieCachingStorage> { Some(self) }
}
```
then use `as_recording_storage` and `as_caching_storage` instead of `downcast_ref::<TrieRecordingStorage>()` and `self.storage.downcast_ref::<TrieCachingStorage>()`